### PR TITLE
Implement GetLastFailure API

### PR DESCRIPTION
**Pull request by Sushisource**
Resolves https://github.com/temporalio/sdk-go/issues/291

Added unit & integ test

### CODE REVIEW

File: internal/interceptors.go [hunk 1]
```diff
@@ -86,6 +87,8 @@ type WorkflowOutboundCallsInterceptor interface {
 	IsReplaying(ctx Context) bool
 	HasLastCompletionResult(ctx Context) bool
 	GetLastCompletionResult(ctx Context, d ...interface{}) error
+	HasLastFailure(ctx Context) bool
```
Comment (mfateev) [inline internal/interceptors.go:90]: I don't think HasLastFailure is needed. We can just return nil from GetLastFailure. It didn't work for the completion result as it required deserialization and didn't return anything.

File: internal/workflow_testsuite.go [hunk 2]
```diff
@@ -32,6 +32,8 @@ import (
 	"testing"
 	"time"
 
+	"go.temporal.io/api/failure/v1"
```
Comment (alexshtin) [inline internal/workflow_testsuite.go:35]: There is our internal convention use `pb` suffix for all api(=proto) imports. So this should be `failurepb`. You have it in other places. Please fix it also here too and everywhere else (if any).

**Review by alexshtin — CHANGES_REQUESTED** [review]

File: test/integration_test.go [hunk 3]
```diff
@@ -748,13 +762,19 @@ func (ts *IntegrationTestSuite) executeWorkflowWithContextAndOption(
...
-	return client.StartWorkflowOptions{
+	var wfOptions = client.StartWorkflowOptions{
 		ID:                       wfID,
 		TaskQueue:                ts.taskQueueName,
 		WorkflowExecutionTimeout: 15 * time.Second,
 		WorkflowTaskTimeout:      time.Second,
 		WorkflowIDReusePolicy:    enumspb.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE,
 	}
+	// REVIEW: Putting this here feels a bit lame, but plumbing arguments down through `executeWorkflow`
+	//  feels like a lot of change that almost all the tests don't need. Thoughts?
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply test/integration_test.go:773]: Yeah, I always remove these after reviews - thanks

File: internal/workflow_testsuite.go (hunk 2 above)
Comment (Sushisource) [reply internal/workflow_testsuite.go:35]: Will do

File: test/integration_test.go [hunk 4]
```diff
@@ -684,6 +684,20 @@ func (ts *IntegrationTestSuite) TestContextPropagator() {
 	}, propagatedValues)
 }
 
+const CronWorkflowID = "test-cron"
+
+func (ts *IntegrationTestSuite) TestCron() {
```
Comment (alexshtin) [inline test/integration_test.go:689]: This sounds like we test cron scheduler here but we don't, right? `TestFailurePropogation` probably a better name here. Workflow itself might have a better name.

File: internal/internal_task_handlers.go [hunk 5]
```diff
@@ -633,6 +633,7 @@ func (wth *workflowTaskHandlerImpl) createWorkflowContext(task *workflowservice.
 		Namespace:                wth.namespace,
 		Attempt:                  attributes.GetAttempt(),
 		lastCompletionResult:     attributes.LastCompletionResult,
+		lastFailure:              attributes.ContinuedFailure,
```
Comment (alexshtin) [inline internal/internal_task_handlers.go:636]: I have one conceptual concern here. We don't expose `Failure` from SDK now. We expose go `error`s. Shouldn't you follow same pattern and convert `Failure` to `error`. There is func for this already. @mfateev, what do you think about it?

File: test/integration_test.go (hunk 4 above)
Comment (Sushisource) [reply test/integration_test.go:689]: Indeed

File: internal/internal_task_handlers.go (hunk 5 above)
Comment (mfateev) [reply internal/internal_task_handlers.go:636]: Yes, I missed it. Let's convert it to the error, failure is internal to the framework.

Comment (Sushisource) [reply internal/internal_task_handlers.go]: Fixed. Thanks guys!

File: internal/interceptors.go [hunk 6]
```diff
@@ -86,6 +86,7 @@ type WorkflowOutboundCallsInterceptor interface {
 	IsReplaying(ctx Context) bool
 	HasLastCompletionResult(ctx Context) bool
 	GetLastCompletionResult(ctx Context, d ...interface{}) error
+	GetLastFailure(ctx Context) error
```
Comment (Sushisource) [inline internal/interceptors.go:89]: @mfateev Wanted to get your thoughts on the name of this method. Was talking with @alexshtin and he made a good point that this isn't really consistent with the rest of the SDK - should probably be `GetLastError`. However, that makes Java/Go somewhat inconsistent with each other.

Should we prefer internal consistency, or a consistency of terms? Perhaps we stay consistent internally for now, and try to be more cross-language consistent in a V2?

Comment (mfateev) [reply internal/interceptors.go:89]: I think GetLastError fits Go SDK better.

File: internal/internal_task_handlers.go [hunk 7]
```diff
@@ -633,6 +633,7 @@ func (wth *workflowTaskHandlerImpl) createWorkflowContext(task *workflowservice.
 		Namespace:                wth.namespace,
 		Attempt:                  attributes.GetAttempt(),
 		lastCompletionResult:     attributes.LastCompletionResult,
+		lastFailure:              convertFailureToError(attributes.ContinuedFailure, wth.dataConverter),
```
Comment (mfateev) [inline internal/internal_task_handlers.go:636]: Let's keep Failure at this level. DataConverter doesn't really belong here.

File: test/integration_test.go [hunk 8]
```diff
@@ -684,6 +684,20 @@ func (ts *IntegrationTestSuite) TestContextPropagator() {
 	}, propagatedValues)
 }
 
+const CronWorkflowID = "test-cron"
+
+func (ts *IntegrationTestSuite) TestFailurePropagation() {
+	var expected int
+	err := ts.executeWorkflow(CronWorkflowID, ts.workflows.CronWorkflow, &expected)
+	// Workflow finishes itself by cancelling
```
Comment (mfateev) [inline test/integration_test.go:692]: It is kind of a mess the way it is implemented. But the whole idea of cancellation is that it comes from outside. The workflow finishing itself by canceling is a misnomer.

Comment (Sushisource) [reply test/integration_test.go:692]: Sure, makes sense. Is there a better way I should do this, or just clarifying the comment?

File: internal/interceptors.go (hunk 6 above)
Comment (Sushisource) [reply internal/interceptors.go:89]: Sounds good. Any thoughts on the overall point?

File: internal/internal_task_handlers.go (hunk 7 above)
Comment (Sushisource) [reply internal/internal_task_handlers.go:636]: Fixed

File: internal/internal_workflow_testsuite.go [hunk 9]
```diff
@@ -2232,6 +2232,10 @@ func (env *testWorkflowEnvironmentImpl) setLastCompletionResult(result interface
 	env.workflowInfo.lastCompletionResult = data
 }
 
+func (env *testWorkflowEnvironmentImpl) setLastError(failure error) {
```
Comment (alexshtin) [inline internal/internal_workflow_testsuite.go:2235]:
```suggestion
func (env *testWorkflowEnvironmentImpl) setLastError(err error) {
```
`err` is a common name for variables of `error` type.

Comment (Sushisource) [reply internal/internal_workflow_testsuite.go]: Thx, fixed.

**Review by alexshtin — APPROVED** [review]

**Pull request merged**